### DISCUSSION TITLE
Example of POST request with Multipart file !

### DISCUSSION THREAD

**Discussion by lilian-delouvy** [usebruno/bruno#1471] [Show and tell]
Hello everyone !

Here is an example of a functionning request to send Multipart files with Bruno.

As an example, I used a really basic Spring boot application with a PostMapping route that receives a MultipartFile object and a String "sampleValue".

Here is the content of the controller, just for the sake of clarity:
```
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;
import org.springframework.web.multipart.MultipartFile;
import java.io.IOException;

@RestController
public class SampleController {

    @GetMapping("/health")
    public ResponseEntity<String> getHealthStatus(){
        return ResponseEntity.ok("OK !");
    }

    @PostMapping("/send-file")
    public ResponseEntity<String> uploadFile(
            @RequestParam("file") MultipartFile multipartFile,
            String sampleValue
    ){
        try{
            String fileName = multipartFile.getOriginalFilename();
            String content = new String(multipartFile.getBytes());
            return ResponseEntity.ok(
                    "Received sample value '" + sampleValue +
                            "' with file name " + fileName +
                            " having the following content : " + content);
        } catch (IOException e) {
            return ResponseEntity
                    .status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body("An error occured during file content read");
        }

    }
}
```

Now, create a Bruno request of POST type with the required url (in my example, "http://localhost:8080/send-file") and set the header "Content-Type": "multipart/form-data" in the relevant tab.
In the "Body" tab, specify the type of file as Multipart. You can also define the parameters, but they will be set in the "Vars" tab, so this is more for documentation purposes.

![image](https://github.com/usebruno/bruno/assets/58565718/316d0172-f8a9-45ff-b5fe-5bd2e68ed39b)

In the "Vars" tab, specify the variables:

![image](https://github.com/usebruno/bruno/assets/58565718/653706ab-d375-48f2-8fc8-dd6cee416d86)

Note: if you're on Windows, you will have to tweak your absolute path a bit. By default, you will have something like "C:\something\somethingelse\yourfile.txt". You will have to either update this to "C:\\something\\somethingelse\\yourfile.txt" or "C:/something/somethingelse/yourfile.txt".

Now, we will create our script:

```
const fs = require('fs');
const FormData = require('form-data');

const attachmentPath = bru.getVar("file");

const buffer = fs.createReadStream(attachmentPath);

const data = new FormData();
var test = req.getBody();
data.append("file", buffer);
data.append("sampleValue", bru.getVar("sampleValue"));

req.setBody(data);
```

We're almost ready ! But by default, Bruno is not allowed to access files on your computer, nor is it allowed to use NPM packages without your consent. We do need file system handling and the "form-data" package though. (Note: don't forget to install "form-data" on your computer for this script to work, with ```npm i -g form-data``` for instance).

**EXIT BRUNO** as the following will **corrupt** your collection otherwise (at least, that's what happened to me on my first try).

Go to your "bruno.json" file that defines your collection (as a reminder, you can find the path where you stored it in the settings of you collection, in the "Info" tab). Update it to look like the following:

```
{
  "version": "1",
  "name": "YOUR_COLLECTION_NAME",
  "type": "collection",
  "scripts": {
    "filesystemAccess": {
      "allow": true
    },
    "moduleWhitelist" : [
      "form-data"
    ]
  }
}
```
This allows Bruno to use "fs" and "form-data".

Save the file, and you can open Bruno again. Your request is ready !

![image](https://github.com/usebruno/bruno/assets/58565718/ebc1527d-0f02-4f5f-a758-a7035c703619)

I hope this helped you :)

**Comment by Terseus** [comment]
> I hope this helped you :)

It did, thanks a lot man :)

I can confirm this method works for a FastAPI application that uses [UploadFile](https://fastapi.tiangolo.com/tutorial/request-files/#uploadfile).

**Comment by whoistobias** [comment]
This thread was extremely helpful to me setting up testing for an s3 compatible file upload using presigned POST. Here's some details with an updated method.

First, You no longer need to modify Bruno json to enable external libraries. Follow the instructions on this page to enable developer mode and use npm packages.
https://docs.usebruno.com/testing/script/external-libraries#external-libraries

After that, it's nearly the same, The main difference that all my form fields come from the API. Also, for me, adding `file` must come last. I'm not sure if that's an s3 thing or specific to the testing setup I was using (Rustfs, an s3 compatible docker container).

```js
const fs = require('fs');
const path = require('path');
const FormData = require('form-data');

const data = new FormData();
const absolutePath = path.resolve(bru.cwd(), 'sample.png');
const buffer = fs.createReadStream(absolutePath);

// Append all presigned fields from the previous request
const uploadFields = bru.getVar('upload_fields') || [];
for (const {key, value} of uploadFields) {
  data.append(key, value);
}

data.append('file', buffer)

req.setBody(data)
```

Thanks again for finding this in the first place. Saved me some headache.
